src/main/java/com/kine/Kine/food/service/FoodService.java

```diff
@@ -0,0 +1,106 @@
+
+        Food food = createFoodDTO.ConvertToFood();
+        // Check if a food item with the same details already exists
+        if (foodRepository.existsByNameAndCaloriesAndProteinsAndCarbsAndFat(food.getName(), food.getCalories(), food.getProteins(), food.getCarbs(), food.getFat())){
+            throw new RessourceAlreadyExistException("Food already exists");
+        }
+        if (isNotCaloriesRelationValid(food)) {
+            throw new InvalidDataException("Invalid data");
+        }
+        return foodRepository.save(food).ConvertToCreateFoodDTO();
+    }
+
+    public FoodDTO updateFood(Long id, UpdateDTO newFoodData) throws RessourceNotFoundException, RessourceAlreadyExistException, InvalidDataException {
+
+        Optional<Food> existingFoodOptional = foodRepository.findFoodById(id);
+
+        if (existingFoodOptional.isEmpty()){
+            throw new RessourceNotFoundException("Food with ID "  + id +  " not found");
+        }
+
+        Food existingFood = existingFoodOptional.get();
+
+        // Update only the fields that are provided in the newFoodData object
+        if (newFoodData.getName().isPresent()) {
+            existingFood.setName(newFoodData.getName().get());
+        }
+        if (newFoodData.getCalories().isPresent()) {
+            existingFood.setCalories(newFoodData.getCalories().get());
+        }
+        if (newFoodData.getProteins().isPresent()) {
+            existingFood.setProteins(newFoodData.getProteins().get());
+        }
+        if (newFoodData.getCarbs().isPresent()) {
+            existingFood.setCarbs(newFoodData.getCarbs().get());
+        }
+        if (newFoodData.getFat().isPresent()) {
+            existingFood.setFat(newFoodData.getFat().get());
+        }
+
+        if (isNotCaloriesRelationValid(existingFood)) {
+            throw new InvalidDataException("Invalid data");
+        }
+        if (foodRepository.existsByNameAndCaloriesAndProteinsAndCarbsAndFat(existingFood.getName(), existingFood.getCalories(), existingFood.getProteins(), existingFood.getCarbs(), existingFood.getFat())){
+            throw new RessourceAlreadyExistException("Food already exists");
+        }
+        Food food = foodRepository.save(existingFood);
+        return new FoodDTO(food.getId(), food.getName(), food.getCalories(), food.getProteins(), food.getCarbs(), food.getFat());
+    }
+
+    public void deleteFoodById(Long id) throws RessourceNotFoundException {
+        Optional<Food> foodOptional = foodRepository.findById(id);
+        if (foodOptional.isPresent()) {
+            foodRepository.deleteById(id);
+        } else {
+            throw new RessourceNotFoundException("Food with ID " + id + " not found");
+        }
+    }
+
+    private boolean isNotCaloriesRelationValid(Food food) {
+        double calculatedCalories = (food.getCarbs() * 4) + (food.getProteins() * 4) + (food.getFat() * 9);
```

Comment:
be worry of the type converstion

Reply:
Maybe also add max values and insure they are not negative

Reply:
Java perform widening / automatic type conversion, the conversion from float to double is done here. Will it be better if I store in a float?

Reply:
You usually run the risk of overflowing but we are never at risk with the possible values. Just keep it a float.